### ISSUE TITLE
Проблема пробития чека (возврата) с наличным типом оплаты и внесением в плагине для ККТ

### ISSUE THREAD

**Issue by durachok1107** [iiko/front.api.sdk#457]
При написании плагина для ККТ возникло несколько проблем, решение которых не удаётся найти.
1) При нажатии кнопки Внесение до плагина как-будто не доходит вызов функции, есть только:
 INFO [ 4] - Get Query Info
 INFO [ 4] - OpenDrawer
Как следствие вызвать что-то на ККТ не имеем возможности
2) При проведении чека или возврата, если среди типов оплаты есть Наличными, по логам видим:
 INFO [ 6] - Get Query Info (не всегда)
 INFO [ 6] - OpenDrawer
Если данного типа оплаты в чеке нет -  в логах наблюдаем набор всех функций и мы можем провести соответствующую операцию.

Подскажите, на это могут влиять какие-то настройки, или может не реализован метод в плагине, который вызывается, но сообщений об ошибках никаких нет. Может можно куда-то посмотреть в логи iiko для понимания этого?

**Comment by durachok1107** [comment]
На примере функции Внесение. Провели Открытие смены+Печать Х-отчета и после сразу внесение, логи плагина закончились на этом, сами функции реализованы сейчас так, для теста. Если код из функции Внесения перенести в Выплату, то он выполняется. Т.е. как-будто до плагина вызов функции даже не доходит. Не знаю как ещё лучше продемонстрировать. Может плагин предоставить, если так будет удобно, проверите у себя как-то вызов метода?
[Логи плагина.txt](https://github.com/iiko/front.api.sdk/files/12172510/default.txt)
[Код функций.txt](https://github.com/iiko/front.api.sdk/files/12172511/default.txt)

**Comment by durachok1107** [comment]
Вот различия по логам плагина, если пробивать чек с наличной оплатой и банковская карта.
[Логи чека банковская карта.txt](https://github.com/iiko/front.api.sdk/files/12173804/default.txt)
[Логи чека наличные.txt](https://github.com/iiko/front.api.sdk/files/12173806/default.txt)
Тут возникла мысль, так как при выборе наличных в приложении стоит оплата demo-profitm.iiko.it, нет какой-то проверки, что такие документы в плагин не отправляются?

**Comment by Bobrov88** [comment]
Можно ли получить обратную связь после отправки логов и кода с нашей стороны? Возможно, что-то ещё порекомендуете. С нашей же стороны, кроме как отсутствия вызова метода DoPayIn, на данный момент других идей нет

**Comment by igorpooh1978** [comment]
https://iiko.github.io/front.api.doc/v6/ru/CashRegisters.html
в самом конце есть FAQ прочитайте и возможно ваша проблема тут же решиться

**Comment by Bobrov88** [comment]
При пробитии чека в iiko выходит окно ожидания и закрывается. При наступлении какого события оно закрывается?

**Comment by vcpp** [comment]
> Может плагин предоставить, если так будет удобно, проверите у себя как-то вызов метода?

Да, я это и прошу. Оставить минимум кода, необходимого для воспроизведения, воспроизвести и затем поделиться исходниками плагина (целиком, а не одной функцией) и логами (всем содержимым папки с логами, а не только логов плагина). Если там есть что-то конфиденциальное, лучше через техподдержку передать.

> Тут возникла мысль, так как при выборе наличных в приложении стоит оплата demo-profitm.iiko.it, нет какой-то проверки, что такие документы в плагин не отправляются?

Да, настройки типа оплаты могут иметь значение. Посмотрите, пожалуйста, какие счета используют ваши внесения/изъятия, а также проверьте следующие галочки:
* «Печатать блюда на чеке» в настройках фискального регистратора
* «Является фискальным» в настройках типа оплаты

**Comment by Bobrov88** [comment]
> > Может плагин предоставить, если так будет удобно, проверите у себя как-то вызов метода?
> 
> Да, я это и прошу. Оставить минимум кода, необходимого для воспроизведения, воспроизвести и затем поделиться исходниками плагина (целиком, а не одной функцией) и логами (всем содержимым папки с логами, а не только логов плагина). Если там есть что-то конфиденциальное, лучше через техподдержку передать.
> 
> > Тут возникла мысль, так как при выборе наличных в приложении стоит оплата demo-profitm.iiko.it, нет какой-то проверки, что такие документы в плагин не отправляются?
> 
> Да, настройки типа оплаты могут иметь значение. Посмотрите, пожалуйста, какие счета используют ваши внесения/изъятия, а также проверьте следующие галочки:
> 
> * «Печатать блюда на чеке» в настройках фискального регистратора
> * «Является фискальным» в настройках типа оплаты

Эти проблемы решены. 

Просим ответить на вчерашнее сообщение.

**Comment by vcpp** [comment]
> При пробитии чека (именно покупка) в iiko выходит окно ожидания и закрывается. При наступлении какого события оно закрывается?

Окно ожидания показывается, когда инициированную пользователем операцию невозможно выполнить мгновенно и нужно подождать. Когда операция завершится, окно ожидания закроется. Примеры длительных операций: синхронное взаимодействие с оборудованием (фискальный регистратор, принтер, весы и т. п.), синхронные обращения к плагинам, синхронные сетевые запросы и пр.

**Comment by Bobrov88** [comment]
> > При пробитии чека (именно покупка) в iiko выходит окно ожидания и закрывается. При наступлении какого события оно закрывается?
> 
> Окно ожидания показывается, когда инициированную пользователем операцию невозможно выполнить мгновенно и нужно подождать. Когда операция завершится, окно ожидания закроется. Примеры длительных операций: синхронное взаимодействие с оборудованием (фискальный регистратор, принтер, весы и т. п.), синхронные обращения к плагинам, синхронные сетевые запросы и пр.

Окно ожидания при пробитии чека продажи закрывается раньше на значительное время. При этом операция ещё не завершена и чек не напечатан. Определить, что вызывает закрытие окна, на данный момент не представляется возможным. Возможно установлен какой-то временной порог либо имеются настройки на этот счёт, так как другие операции, повторюсь, завершаются параллельно с закрытием окна ожидания.

**Comment by Bobrov88** [comment]
> > При пробитии чека (именно покупка) в iiko выходит окно ожидания и закрывается. При наступлении какого события оно закрывается?
> 
> Окно ожидания показывается, когда инициированную пользователем операцию невозможно выполнить мгновенно и нужно подождать. Когда операция завершится, окно ожидания закроется. Примеры длительных операций: синхронное взаимодействие с оборудованием (фискальный регистратор, принтер, весы и т. п.), синхронные обращения к плагинам, синхронные сетевые запросы и пр.

В качестве эксперимента в метод openDrawer была добавлена искусственная задержка выполнения кода. Окно ожидания при "Внесение наличных" зависло как и предполагалось, то есть в данном случае iiko front ждёт завершения операции. При оплате наличкой же - окно ожидания закрылось, как и без добавления искусственной задержки, что говорит, что в данном случае iiko front не дожидается выполнения операции. Прошу Вас проверить этот момент на стороне iiko
